### PR TITLE
implement new HTTP request handling framework

### PR DESCRIPTION
**Pull request by thrau**
## Summary

This PR adds an HTTP request handling framework based on the chain of responsibility pattern, where requests are processed by Handlers that are part of a Handler Chain. This framework replaces the edge proxy, generic proxy, and http2_server.  It provides a clean interface for serving a Gateway (a request -response interface) through ASGI or WSGI.

The behavior of LocalStack is implemented in the `LocalstackGateway`, and can be served directly through hypercorn (try `python -m localstack.aws.app` for serving it directly without bootstrapping)

It can be enabled by setting `LEGACY_EDGE_PROXY=0`.

The current version only works for community. Pro integration will be done in a second iteration (pro test failures are expected). A proper developer documentation will be provided then.

## PRs

The following PRs were the result of the changes developed in this PR.
* #5478 
* #5734
* #5876
* #5970 
* #6040 
* #6006
* #6012
* #6190 
* #6191
* #6192 
* ... and many other fixes and extensions in the asf code, many of which @alexrashed has contributed

## Tasks
- [x] Implement new HTTP request handling (while not breaking the previous functionality).
- [x] Migrate current features of the edge proxy (service detection,...) into separate handlers.
- [x] Fix all community integration tests with the new request handling enabled.
- [x] Hide the new functionality behind a feature flag (opt-in).
- [x] Make the PR smaller by pulling out sub-PRs for dependent changes in master
- [x] Add documentation for the handler chain framework
- [x] Final polishing iteration
- [x] PR review
- [x] change default value of `LEGACY_EDGE_PROXY` back to 1 and run one last test

### CODE REVIEW

File: localstack/aws/chain.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                    self.error = e
+                if self.stop_on_error:
+                    self.stopped = True
+
+                # call exception handlers safely
+                self._call_exception_handlers(e, response)
+
+            # decide next step
+            if self.error:
+                raise self.error
```
Comment (whummer) [inline localstack/aws/chain.py:70]: Over time we could consider introducing some sort of a finally-handler to reliably perform cleanup logic. If we raise the error here, then the response handlers would not get called in line 77, and the request context in `quart.globals._request_ctx_stack` would potentially remain on the stack. 

Should not be an issue for now (looks like none of the handlers are calling `.throw()` yet, and since we have `raise_on_error=False` by default), but something we can consider for the future iterations. 👍

**Review by whummer — APPROVED** [review]
Wooah, fantastic set of changes 🚀🚀 - kudos @thrau and @alexrashed for pushing this major refactoring over the line and enabling the next evolution of our HTTP request handling! 🎉 

The new handler chain is soo much cleaner (and easily extensible!), and the WSGI/ASGI bridge you've put in place is phenomenal - now we can get rid of quart and serve the gateway directly in hypercorn (or even nginx directly!). Incredibly excited to get this in! 🥳 

Added a few nitpick comments and questions for clarification - nothing major that should hold back the merge.. 👍

File: localstack/aws/gateway.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    response_handlers: List[Handler]
+    exception_handlers: List[ExceptionHandler]
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.request_handlers = list()
+        self.response_handlers = list()
+        self.exception_handlers = list()
+
+    def new_chain(self, request: Request, response: Response) -> HandlerChain:
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/gateway.py:27]: yes that was the reasoning. but you're right that we can introduce that easily at some other point should we need it.

File: localstack/aws/handlers/cors.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            not config.DISABLE_CORS_CHECKS
+            and self.should_enforce_self_managed_service(context)
+            and not self.is_cors_origin_allowed(context.request.headers)
+        ):
+            LOG.info(
+                "Blocked CORS request from forbidden origin %s",
+                context.request.headers.get("origin") or context.request.headers.get("referer"),
+            )
+            response.status_code = 403
+            chain.terminate()
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/handlers/cors.py:110]: /cc @alexrashed

File: localstack/aws/handlers/fallback.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    """
+
+    def __call__(
+        self,
+        chain: HandlerChain,
+        exception: Exception,
+        context: RequestContext,
+        response: Response,
+    ):
+        if response.response:
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/handlers/fallback.py:26]: The `response` attribute of `Response` is the underlying container for `data`. `response` is always an iterable of bytes (or strings that are later encoded), whereas `data` is a dynamic property that is the rendered output of `response`.

that means the following things:
* `response.response` can never be None. For an empty response, it will be an empty list. `response.response is not None` will therefore always return true
* in the case that `response.response` contains a generator (in the kinesis streaming case for example) calling `response.data` will block until the entire generator has been consumed, so calling `if request.data` is in general not safe.

for setting data, it means the following:
`data` has getters and setters which operate on strings and bytes. calling `response.data = "foobar"` is the same as calling `response.set_data("foobar")`, which will set `self.response = ["foobar".encode('utf-8')]` (simplified). and also update the Content-Length headers appropriately. so whenever we know that something is a string or bytes, we can safely set `response.data = ...`, whereas if we're not sure whether something may be an iterator, it's better to use `response.set_response(...)` because that will do that check for us (at the cost of doing the check), which will again fall back to `response.data = ...` if the input is a string or bytes.

> Could there be a case here where response.response is falsy, but response.data has already been set (and should be returned)?

not that i can think of. if you set `response.data = b""` (to indicate a deliberately empty response), then `response.response` will be `[b""]` which will be truthy (which is btw exactly the case we are guarding against).
unless you set `response.response` to something very weird that overwrites `__bool__` then yes i guess you could run into that problem. but as you said that doesn't seem very likely to me.

so in summary:
* `response.response` may be set to an iterator -> `response.response` will be truthy
* we cant do `if response.data` checks because it would potentially consume the underlying chunked response stream
* an empty response (`data = ""` which is equivalent to `set_response("")`) -> `response` will be automatically truthy
* `data` never set -> `response` will be falsy

File: localstack/aws/handlers/service.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        service = determine_aws_service_name(context.request)
+
+        if not service:
+            return
+
+        context.service = self.get_service_model(service)
+        headers = context.request.headers
+        headers["x-localstack-tgt-api"] = service
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/handlers/service.py:37]: fair point. tbh i'm not sure we even need this header anymore. i put a TODO in there to remove it. i would try to avoid importing legacy code in the new modules.

File: localstack/aws/handlers/routes.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+LOG = logging.getLogger(__name__)
+
+
+class RouterHandler(Handler):
+    """
+    Adapter to serve a Router as a Handler.
+    """
+
+    resources: Router
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/handlers/routes.py:18]: good catch!

File: localstack/aws/handlers/service.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        if not context.service:
+            LOG.debug("no service set in context, skipping request parsing")
+            return
+
+        return self.parse_and_enrich(context)
+
+    def get_parser(self, service: ServiceModel):
+        name = service.service_name
+
+        if name in self.parsers:
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/handlers/service.py:66]: agreed. the parsers are stateless so it's fine to just create a new one and overwrite the old one.

File: localstack/aws/handlers/service_plugin.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                return
+
+            service_name = context.service.service_name
+            if not self.service_manager.exists(service_name):
+                raise NotImplementedError
+
+            service_plugin: Service = self.service_manager.require(service_name)
+
+            if isinstance(service_plugin, Service):
+                if type(service_plugin.listener) == AwsApiListener:
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/aws/handlers/service_plugin.py:59]: i know this looks weird here, but it's very explicitly not an `isinstance` check. 

the `AwsApiListener` is completely unnecessary with the new gateway. it currently serves as the glue between the ASF Skeleton and the edge proxy. moving forward, the `AwsApiListener` will be removed. you could instead pass the instance of `provider` directly to the `Service(listener=...)`, but this would break backwards compatibility.
so this code here extracts from the `AwsApiListener` the `skeleton` and adds it to the router directly. we can do that because we know that `AwsApiListener` will do nothing else and not invoking it will have no effect. we cannot make that assumption for other listeners that extend `AwsApiListener` (like `DynamoDBApiListener` which still has the `return_response` method), so we need to invoke them through the `LegacyPluginHandler`.

we can start cleaning this up when all services have been migrated to ASF, and the legacy edge proxy has been removed after the deprecation phase.

File: localstack/aws/chain.py (hunk 1 above)
Comment (thrau) [reply localstack/aws/chain.py:70]: yea this is also a construct that was really introduced in the interest of generality. definitely not something that you would use in a normal production setting or with the `LocalstackGateway`. but it may come in handy for testing/debugging. other handler chains may want to re-raise errors immediately. :shrug: let's see whether we really need it

File: localstack/http/request.py [hunk 9]
```diff
@@ -188,3 +189,37 @@ def get_full_raw_path(request) -> str:
...
+    could make . This is a useful method to have when we are proxying requests, i.e., create outgoing requests from
+    incoming requests that were subject to previous parsing.
+
+    TODO: this construct is not great and will definitely be a source of trouble. but something like this will
+      inevitably become the the basis of proxying werkzeug requests. The alternative is to build our own request object
+      that memoizes the original payload before parsing.
+    """
+    data = request.data
+
+    if request.method != "POST":
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/http/request.py:207]: oh man totally missed that. yes i guess that's possible. it probably would be better to check for `request.want_form_data_parsed()`. i'll take a note for the next pr

File: localstack/aws/handlers/fallback.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (whummer) [reply localstack/aws/handlers/fallback.py]: Makes perfect sense (should have taken a more careful look at the `data` property). Thanks for the detailed explanation! 👍

File: localstack/aws/chain.py (hunk 1 above)
Comment (whummer) [reply localstack/aws/chain.py:70]: 👌
